### ISSUE TITLE
repair did not send row after replication factor increase

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#11684]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/106/testReport/ttl_test/TestDistributedTTL/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split004___test_ttl_is_respected_on_repair/
Scylla version 060dda8e00b7 with build-id b4e08d869feb2fecb04c6ea45eb8946cade10c70
```
self = <ttl_test.TestDistributedTTL object at 0x7f78a211a430>

    @pytest.mark.next_gating
    @pytest.mark.dtest_debug
    def test_ttl_is_respected_on_repair(self):
        """ Test that ttl is respected on repair """
    
        self.prepare()
        self.session1.execute("""
            ALTER KEYSPACE ks WITH REPLICATION =
            {'class' : 'SimpleStrategy', 'replication_factor' : 1};
        """)
        self.session1.execute("""
            INSERT INTO ttl_table (key, col1) VALUES (1, 1) USING TTL 5;
        """)
        self.session1.execute("""
            INSERT INTO ttl_table (key, col1) VALUES (2, 2) USING TTL 1000;
        """)
    
        assert_all(
            self.session1,
            "SELECT * FROM ttl_table;",
            [[1, 1, None, None], [2, 2, None, None]]
        )
        time.sleep(7)
        self.node1.stop()
        session2 = self.patient_exclusive_cql_connection(self.node2)
        session2.execute("USE ks;")
        assert_unavailable(session2.execute, "SELECT * FROM ttl_table;")
        self.node1.start(wait_for_binary_proto=True)
        self.session1 = self.patient_exclusive_cql_connection(self.node1)
        self.session1.execute("USE ks;")
        self.session1.execute("""
            ALTER KEYSPACE ks WITH REPLICATION =
            {'class' : 'SimpleStrategy', 'replication_factor' : 2};
        """)
        self.node1.repair(['ks'])
        ttl_start = time.time()
        ttl_session1 = self.session1.execute('SELECT ttl(col1) FROM ttl_table;')
        self.node1.stop()
    
>       assert_row_count(session2, 'ttl_table', 1)

ttl_test.py:778: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tools/retrying.py:43: in inner
    return func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <cassandra.cluster.Session object at 0x7f78a1438790>
table_name = 'ttl_table', expected = 1, consistency_level = 1, num_attempts = 1
timeout = None

    @retrying(num_attempts=1, sleep_time=10)
    def assert_row_count(session, table_name, expected, consistency_level=ConsistencyLevel.ONE, num_attempts=1,
                         timeout=None):
        """
        Function to validate the row count expected in table_name
        @param session Session to use
        @param table_name Name of the table to query
        @param expected Number of rows expected to be in table
        @param num_attempts defines how many time try to assert data in case failure. Used in retrying decorator
        @param timeout
    
        Examples:
        assert_row_count(self.session1, 'ttl_table', 1)
        """
        from tools.data import run_query_with_data_processing  # to avoid cyclic dependency
    
        query = "SELECT count(*) FROM {}".format(table_name)
        count = run_query_with_data_processing(session, query, consistency_level=consistency_level, session_timeout=timeout)
        if isinstance(count, list):
            count = count[0][0]
>       assert count == expected, "Expected a row count of {} in table '{}', but got {}".format(
            expected, table_name, count)
E       AssertionError: Expected a row count of 1 in table 'ttl_table', but got 0

tools/assertions.py:257: AssertionError
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/106/artifact/logs-full.release.004/1664708251830_ttl_test.py%3A%3ATestDistributedTTL%3A%3Atest_ttl_is_respected_on_repair/node1.log
```
Scylla version 5.2.0~dev-0.20221002.060dda8e00b7 with build-id b4e08d869feb2fecb04c6ea45eb8946cade10c70 starting ...
...
INFO  2022-10-02 10:57:22,437 [shard 0] schema_tables - Altering keyspace ks
INFO  2022-10-02 10:57:22,443 [shard 0] schema_tables - Schema version changed to 0aa0ebcf-0ba3-3edf-976c-4ea38651bd7a
INFO  2022-10-02 10:57:23,966 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: starting user-requested repair for keyspace ks, repair id 1, options {{ trace -> false}, { jobThreads -> 1}, { incremental -> false}, { parallelism -> parallel}, { primaryRange -> false}}
INFO  2022-10-02 10:57:23,966 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Skipped sending repair_flush_hints_batchlog to nodes=[127.0.24.2, 127.0.24.1]
INFO  2022-10-02 10:57:23,967 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Started to repair 1 out of 1 tables in keyspace=ks, table=ttl_table, table_id=e6815850-4240-11ed-8fee-2d4913f5b00e, repair_reason=repair
INFO  2022-10-02 10:57:23,967 [shard 1] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Started to repair 1 out of 1 tables in keyspace=ks, table=ttl_table, table_id=e6815850-4240-11ed-8fee-2d4913f5b00e, repair_reason=repair
WARN  2022-10-02 10:57:23,969 [shard 1] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Repair 4 out of 513 ranges,  shard=1, keyspace=ks, table={ttl_table}, range=(-9193877161816225917, -9191332335124261509], peers={}, live_peers={}, status=skipped_no_followers
...
WARN  2022-10-02 10:57:24,078 [shard 1] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Repair 511 out of 513 ranges,  shard=1, keyspace=ks, table={ttl_table}, range=(9124849779871422718, 9192934245907542968], peers={}, live_peers={}, status=skipped_no_followers
INFO  2022-10-02 10:57:24,079 [shard 1] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: shard 1 stats: repair_reason=repair, keyspace=ks, tables={ttl_table}, ranges_nr=513, round_nr=257, round_nr_fast_path_already_synced=257, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=1285, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.11183742 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={{127.0.24.1, 0}, {127.0.24.2, 0}}, row_from_disk_nr={{127.0.24.1, 0}, {127.0.24.2, 0}}, row_from_disk_bytes_per_sec={{127.0.24.1, 0}, {127.0.24.2, 0}} MiB/s, row_from_disk_rows_per_sec={{127.0.24.1, 0}, {127.0.24.2, 0}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
INFO  2022-10-02 10:57:24,079 [shard 1] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: shard 1 completed successfully, keyspace=ks
INFO  2022-10-02 10:57:24,148 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: shard 0 stats: repair_reason=repair, keyspace=ks, tables={ttl_table}, ranges_nr=513, round_nr=514, round_nr_fast_path_already_synced=513, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=1, rpc_call_nr=2568, tx_hashes_nr=0, rx_hashes_nr=1, duration=0.18178925 seconds, tx_row_nr=0, rx_row_nr=1, tx_row_bytes=0, rx_row_bytes=172, row_from_disk_bytes={{127.0.24.1, 0}, {127.0.24.2, 172}}, row_from_disk_nr={{127.0.24.1, 0}, {127.0.24.2, 1}}, row_from_disk_bytes_per_sec={{127.0.24.1, 0}, {127.0.24.2, 0.00090232}} MiB/s, row_from_disk_rows_per_sec={{127.0.24.1, 0}, {127.0.24.2, 5.50088}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={{127.0.24.2, 1}}
INFO  2022-10-02 10:57:24,148 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: shard 0 completed successfully, keyspace=ks
INFO  2022-10-02 10:57:24,148 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Started to shutdown off-strategy compaction updater
INFO  2022-10-02 10:57:24,148 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: Finished to shutdown off-strategy compaction updater
INFO  2022-10-02 10:57:24,148 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: completed successfully
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/106/artifact/logs-full.release.004/1664708251830_ttl_test.py%3A%3ATestDistributedTTL%3A%3Atest_ttl_is_respected_on_repair/node2.log
```
INFO  2022-10-02 10:57:22,437 [shard 0] schema_tables - Altering keyspace ks
INFO  2022-10-02 10:57:22,443 [shard 0] schema_tables - Schema version changed to 0aa0ebcf-0ba3-3edf-976c-4ea38651bd7a
INFO  2022-10-02 10:57:25,497 [shard 0] gossip - Got shutdown message from 127.0.24.1, received_generation=1664708217, local_generation=1664708217
```

**Comment by bhalevy** [comment]
@asias why wasn't the live row repaired?

**Comment by bhalevy** [comment]
Cc @eliransin

**Comment by eliransin** [comment]
@asias this looks like a correctness issue please handle with high priority.

**Comment by asias** [comment]
@bhalevy I assume this is a recent regression. Do you know the last good version?

**Comment by asias** [comment]
@bhalevy  can you confirm the dtest and ccm version used in the report?

![image](https://user-images.githubusercontent.com/320076/196082054-e6601767-42fb-47e0-8b69-067122e2bd26.png)

**Comment by asias** [comment]
Looking at the logs 

```
INFO  2022-10-02 10:57:24,079 [shard 1] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: shard 1 stats: repair_reason=repair, keyspace=ks, tables={ttl_table}, ranges_nr=513, round_nr=257, round_nr_fast_path_already_synced=257, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=1285, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.11183742 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={{127.0.24.1, 0}, {127.0.24.2, 0}}, row_from_disk_nr={{127.0.24.1, 0}, {127.0.24.2, 0}}, row_from_disk_bytes_per_sec={{127.0.24.1, 0}, {127.0.24.2, 0}} MiB/s, row_from_disk_rows_per_sec={{127.0.24.1, 0}, {127.0.24.2, 0}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
INFO  2022-10-02 10:57:24,148 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: shard 0 stats: repair_reason=repair, keyspace=ks, tables={ttl_table}, ranges_nr=513, round_nr=514, round_nr_fast_path_already_synced=513, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=1, rpc_call_nr=2568, tx_hashes_nr=0, rx_hashes_nr=1, duration=0.18178925 seconds, tx_row_nr=0, rx_row_nr=1, tx_row_bytes=0, rx_row_bytes=172, row_from_disk_bytes={{127.0.24.1, 0}, {127.0.24.2, 172}}, row_from_disk_nr={{127.0.24.1, 0}, {127.0.24.2, 1}}, row_from_disk_bytes_per_sec={{127.0.24.1, 0}, {127.0.24.2, 0.00090232}} MiB/s, row_from_disk_rows_per_sec={{127.0.24.1, 0}, {127.0.24.2, 5.50088}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={{127.0.24.2, 1}}
```

So,  n1 initially did not contain  the row(2,2), n2 contained row(2,2), the repair above showed n1 got the row(2,2) from n2. 

But the test complained n2, did not contain the row(2,2) on n2
```
 assert_row_count(session2, 'ttl_table', 1)
```

This sounds like a bug in the query where the row(2,2) with ttl 2000 was not returned in the query to n2.

**Comment by bhalevy** [comment]
Cc @tgrabiec @denesb any recent changes to the query logic that could explain the above?

We should bisect the branch and pinpoint the patch that caused this regression.

**Comment by denesb** [comment]
I see no flushes, maybe they new compact-on-read gets rid of the ttl'd data from memtable.

**Comment by tgrabiec** [comment]
Isn't repair writing to sstables directly?

**Comment by denesb** [comment]
I understood that regular reads don't show the row either. If the row was collected before repair would start then repair will not have a chance of sending it.

**Comment by tgrabiec** [comment]
The row shouldn't be collected, it has ttl of 1000.

**Comment by tgrabiec** [comment]
I don't have an idea what could be wrong, we need to reproduce it.

**Comment by denesb** [comment]
> I don't have an idea what could be wrong, we need to reproduce it.

Yeah, I was also just shooting from the hip and this was the first thing that came to my mind.

**Comment by asias** [comment]
> Isn't repair writing to sstables directly?

yes

**Comment by asias** [comment]
There are only two rows in the tests

row(1,1) with ttl 5
row(2,2) with ttl 1000

Initially,  the rf = 1, then rf is increased to 2 and repair was performed, repair is supposed to bring row(2,2) with ttl 1000 to the node that does not own it when RF =1, repair did that form the log (n1 did not have row(2,2), n2 had row(2,2) when RF =1)

However, the strange thing is that the test complained n2 did have row(2,2) according to the query.

**Comment by bhalevy** [comment]
@Jadw1 since the issue happens in `SELECT count(*)`, please look into this issue

**Comment by bhalevy** [comment]
I suspect this might be related to 5a30f9b7894abbeb196d179294f9320cc53534ba

**Comment by bhalevy** [comment]
Still happens from time to time, e.g. https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/132/testReport/junit/ttl_test/TestDistributedTTL/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split004___test_ttl_is_respected_on_repair/

**Comment by bhalevy** [comment]
ping @Jadw1 please update status of your investigation

**Comment by Jadw1** [comment]
I'm not able to reproduce this. 
I'm running release build on https://github.com/scylladb/scylladb/commit/060dda8e00b74fbb0eebc46ca08102250686b780 commit and the test passes.
Tried multiple times with both `enable_parallelized_aggregation` set to true and false (false disables changes of my https://github.com/scylladb/scylladb/commit/5a30f9b7894abbeb196d179294f9320cc53534ba commit).

The test is executed this way:
```
export SCYLLA_EXT_OPTS="--smp 2"
pytest --cassandra-dir=/home/jadw1/src/scylla/build/release ttl_test.py::TestDistributedTTL::test_ttl_is_respected_on_repair
```
Am I missing something to make the test fail like those on jenkins?

**Comment by DoronArazii** [comment]
@Jadw1 please keep us posted with the progress

**Comment by DoronArazii** [comment]
@eliransin please update on the status here

**Comment by fruch** [comment]
@Jadw1 it happened today on next:

https://jenkins.scylladb.com/job/scylla-master/job/next/5590/testReport/junit/ttl_test/TestDistributedTTL/Sanity_Tests__dtest____dtest___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_/

and on enterprise daily run:

https://jenkins.scylladb.com/job/scylla-enterprise/job/dtest-daily-release/108/testReport/ttl_test/TestDistributedTTL/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split003___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_/

**Comment by eliransin** [comment]
> @Jadw1 it happened today on next:
> 
> https://jenkins.scylladb.com/job/scylla-master/job/next/5590/testReport/junit/ttl_test/TestDistributedTTL/Sanity_Tests__dtest____dtest___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_/

The failed test is with aggregate queries disabled. Will try to make sure what is going on.

**Comment by mykaul** [comment]
@eliransin - any updates?

**Comment by nyh** [comment]
Seen again today in https://jenkins.scylladb.com/job/scylla-master/job/next/5654/

**Comment by nyh** [comment]
Also seen in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4176/. Failed both parallel_aggregates_disabled and parallel_aggregates_enabled versions.

**Comment by michoecho** [comment]
Also seen in https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4542/

**Comment by Jadw1** [comment]
@mykaul I started looking into this yestarday, but I still cannot reproduce this locally. @eliransin had an idea to print sstable before and after the altering RF and repair, so it might give us any more information. This PR adds this prints https://github.com/scylladb/scylla-dtest/pull/3046

**Comment by eliransin** [comment]
https://github.com/scylladb/scylla-dtest/pull/3046 is merged. Now we need to wait for it to fail.
Just to summarize what we should do after it failes:
We need to look at the logs and we will be able to tell, what went wrong (I assume the the records prior to the repair
are good even though we will see them in the log):
1. If the sstables doesn't contain the right records after the RF change and repair, this is a repair issue.
2. If the records contains the right records it is either a problem with the reconciliation or a problem with our expectation
from the query (for example it does return the TTL's data but this is what was supposed to happen).

**Comment by eliransin** [comment]
We hit this again on the enterprise run:
The added logs:
```
10:55:11,083 540     ttl_test                       DEBUG    ttl_test.py         :790  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: SStable dump after writes with RF = 1
10:55:11,084 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=flush wait=True timeout=None
10:55:11,940 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=cfstats wait=True timeout=60
10:55:15,022 540     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:55:04.078722Z", "ttl" : 1000, "expires_at" : "2023-02-26T11:11:44Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
10:55:15,022 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=flush wait=True timeout=None
10:55:15,829 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=cfstats wait=True timeout=60
10:55:19,008 540     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:55:04.077746Z", "ttl" : 5, "expires_at" : "2023-02-26T10:55:09Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
```

```
10:55:49,380 540     ttl_test                       DEBUG    ttl_test.py         :805  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: SStable dump after alter RF to 2 and repair
10:55:49,381 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=flush wait=True timeout=None
10:55:50,145 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=cfstats wait=True timeout=60
10:55:54,285 540     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:55:04.078722Z", "ttl" : 1000, "expires_at" : "2023-02-26T11:11:44Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:55:04.077746Z", "ttl" : 5, "expires_at" : "2023-02-26T10:55:09Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
10:55:54,286 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=flush wait=True timeout=None
10:55:55,019 540     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=cfstats wait=True timeout=60
10:55:58,835 540     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:55:04.077746Z", "ttl" : 5, "expires_at" : "2023-02-26T10:55:09Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:55:04.078722Z", "ttl" : 1000, "expires_at" : "2023-02-26T11:11:44Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
```

**Comment by eliransin** [comment]
Happened on enterprise - and the conclusion is that something is probably wrong with repair (@asias), here are the logs of the sstables before and after (thanks @Jadw1):
Before repair:
```
test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: SStable dump after writes with RF = 1
10:54:52,439 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=flush wait=True timeout=None
10:54:53,311 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=cfstats wait=True timeout=60
10:54:56,512 552     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:54:45.432079Z", "ttl" : 1000, "expires_at" : "2023-02-26T11:11:25Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
10:54:56,513 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=flush wait=True timeout=None
test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=cfstats wait=True timeout=60
10:55:00,729 552     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:54:45.430554Z", "ttl" : 5, "expires_at" : "2023-02-26T10:54:50Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
```

```
10:55:29,316 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=repair ks wait=True timeout=None
```

After repair:
```
10:55:31,605 552     ttl_test                       DEBUG    ttl_test.py         :805  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: SStable dump after alter RF to 2 and repair
10:55:31,606 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=flush wait=True timeout=None
10:55:32,373 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=cfstats wait=True timeout=60
10:55:36,565 552     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:54:45.432079Z", "ttl" : 1000, "expires_at" : "2023-02-26T11:11:25Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:54:45.430554Z", "ttl" : 5, "expires_at" : "2023-02-26T10:54:50Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
10:55:36,566 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=flush wait=True timeout=None
10:55:37,302 552     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=cfstats wait=True timeout=60
10:55:40,037 552     ttl_test                       DEBUG    ttl_test.py         :656  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-02-26T10:54:45.430554Z", "ttl" : 5, "expires_at" : "2023-02-26T10:54:50Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
```

**Comment by eliransin** [comment]
From the logs above it can be seen that for some reason node2's sstables hasn't been repaired correctly.
@asias please have a look.
The link for the job is: https://jenkins.scylladb.com/job/scylla-enterprise/job/gating-dtest-release/15/

**Comment by michoecho** [comment]
> From the logs above it can be seen that for some reason node2's sstables hasn't been repaired correctly.

That is a given, because the test only runs repair on node1, but not on node2...

**Comment by michoecho** [comment]
> > From the logs above it can be seen that for some reason node2's sstables hasn't been repaired correctly.
> 
> That is a given, because the test only runs repair on node1, but not on node2...

Or am I misunderstanding how repair works? Is `nodetool repair` also supposed to affect nodes other than the one it's ran on?

**Comment by denesb** [comment]
On Mon, 2023-02-27 at 03:08 -0800, Michał Chojnowski wrote:
> > > > From the logs above it can be seen that for some reason node2's
> > > > sstables hasn't been repaired correctly.
> > > That is a given, because the test only runs repair on node1, but
> > > not on node2...
> > Or am I misunderstanding how repair works? Is nodetool repair also
> > supposed to affect nodes other than the one it's ran on?
> Nevermind, I think I was misunderstanding indeed, and it's supposed
> to affect both nodes. Removing my comments because they are spam.

Repair affects both nodes, but do note that a repair started on nodeX
will only repair ranges owned by said node.
Furthermore, if I recall correctly, there is a switch (or is this the
default), which when used causes repair to only repair ranges that
nodeX is the primary replica for.

**Comment by eliransin** [comment]
> Repair affects both nodes, but do note that a repair started on nodeX will only repair ranges owned by said node. Furthermore, if I recall correctly, there is a switch (or is this the default), which when used causes repair to only repair ranges that nodeX is the primary replica for.

Just to make it clear (@denesb correct me if I am talking nonsense), this is not what is suspected to happen here because:
1. The replication factor is 2 and number of nodes is 2 so a repair from either node should repair everithing.
2. If this primary replica switch is on, this would have been a constant failure, but this test is flaky AFAICT.

**Comment by nyh** [comment]
Here's my wild guess trying to explain why the test can be flaky:

Although the test always writes the same two partition keys, the **token ring** is chosen _randomly_ on each run, so the location of the two partitions on the two nodes is random. In about 25% of the runs, the two partitions will be on node 2.

Now, imagine that repair had a bug (@asias this is just a wild guess) where if a repair slave doesn't have **any** rows in some range it skips this range instead of correctly talking to the master and fetching its missing rows - then in 25% of the runs, the node with the missing data will not get it. Then, half of those runs will do the read (with CL=1) from the node with the missing data, and return nothing.

My wild guess above yields a test failure in exactly 1/8th of the runs. Is this number (1/8th) consistent with the measured failure rate of this test?

P.S. It is very likely that my wild guess is wrong. But at least I tried ;-)

**Comment by mykaul** [comment]
> My wild guess above yields a test failure in exactly 1/8th of the runs. Is this number (1/8th) consistent with the measured failure rate of this test?
> 
> P.S. It is very likely that my wild guess is wrong. But at least I tried ;-)

So running the tests 100 or 1000 times should kinda help prove that?

**Comment by nyh** [comment]
@mykaul did you ever try to run a dtest 100 times? :-)

**Comment by mykaul** [comment]
> @mykaul did you ever try to run a dtest 100 times? :-)

So should we port it to test.py?

**Comment by michoecho** [comment]
Why is shard 1 on the repair master skipping half of the ranges? (Shard 0 isn't skipping any).
```
INFO  2023-02-26 10:55:30,177 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Started to repair 1 out of 1 tables in keyspace=ks, table=ttl_table, table_id=f9f66c80-b5c3-11ed-b97d-87bf283c710a, repair_reason=repair
WARN  2023-02-26 10:55:30,179 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 4 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-9150195355466686312, -9147322006302888805], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,179 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 5 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-9147322006302888805, -9146362554640448859], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,179 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 6 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-9146362554640448859, -9116303322925318316], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,180 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 9 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-9050042585141535063, -8944657511567347214], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,181 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 13 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-8810844902892324494, -8790377230693553660], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,181 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 14 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-8790377230693553660, -8554330814231781946], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,182 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 16 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-8527656670300930157, -8522507567661944136], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,183 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 19 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-8520755323257219762, -8484506361723539852], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,183 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 21 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-8473682388204627114, -8389883033547621354], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,185 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 27 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-8137633573694440923, -8111253278693074483], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,186 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 29 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-7983782515883722544, -7968883374637446773], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,187 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 32 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-7870276435268461638, -7861072158131330014], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,187 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 33 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-7861072158131330014, -7856188196635790444], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,187 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 34 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-7856188196635790444, -7756637851727750179], peers={}, live_peers={}, status=skipped_no_followers
WARN  2023-02-26 10:55:30,187 [shard 1] repair - repair[e0fcf424-827d-4f6b-8e99-f28caa405611]: Repair 35 out of 513 ranges,  shard=0, keyspace=ks, table={ttl_table}, range=(-7756637851727750179, -7727725288617085281], peers={}, live_peers={}, status=skipped_no_followers
...
```

**Comment by nyh** [comment]
@michoecho good lead. I have a new guess:

I see the test doesn't wait at all between changing the RF and starting the repair.
I think it is possible that one of the nodes doesn't yet realize that with have a new ring, or, alternatively, one node is not aware that the other node is already aware of the new ownerships, when the repair is started, and this can cause these "no followers" problems (and of course a broken repair).
@asias does this make sense? 

If I'm right, adding a long-enough sleep between the change of RF and the start of the repair can make this error go away.

**Comment by bhalevy** [comment]
> @michoecho good lead. I have a new guess:
> 
> I see the test doesn't wait at all between changing the RF and starting the repair. I think it is possible that one of the nodes doesn't yet realize that with have a new ring, or, alternatively, one node is not aware that the other node is already aware of the new ownerships, when the repair is started, and this can cause these "no followers" problems (and of course a broken repair). @asias does this make sense?
> 
> If I'm right, adding a long-enough sleep between the change of RF and the start of the repair can make this error go away.

In the long run changing RF should be executed as a topology change over raft.
During rf change, quorum writes should make sure there's quorum on both the existing replica set and the new one.
Note also that a failure to increase RF should trigger cleanup of the provisional copies that were reverted when the operation is aborted.

**Comment by nyh** [comment]
> In the long run changing RF should be executed as a topology change over raft. 

Let's discuss what we have today in this test, not the long run. I'm at a loss what we have today. Are you suggesting that the "ALTER KEYSPACE" operation doesn't return immediately but rather waits for all nodes to agree on the new ring, and know that they all agree? Maybe that's true, but I was raising the possibility that it isn't. Maybe @asias remembers how this works.

> During rf change, quorum writes should make sure there's quorum on both the existing replica set and the new one. 

This test isn't doing writes during an rf change so this is irrelevant. What I see this test doing (please correct me if I'm wrong) is to change RF with a "ALTER KEYSPACE" and then immediately start a repair operation.

> Note also that a failure to increase RF should trigger cleanup of the provisional copies that were reverted when the operation is aborted.

There isn't any failure to increase RF here. Increasing RF is a trivial operation and it shouldn't (and doesn't) fail. It just takes a small, but non-zero amount of time and I'm not sure (?) if the ALTER KEYSPACE call can't return before everything is really finished and all the nodes agree on the new ring.

**Comment by michoecho** [comment]
> @michoecho good lead. I have a new guess:
> 
> I see the test doesn't wait at all between changing the RF and starting the repair. I think it is possible that one of the nodes doesn't yet realize that with have a new ring, or, alternatively, one node is not aware that the other node is already aware of the new ownerships, when the repair is started, and this can cause these "no followers" problems (and of course a broken repair). @asias does this make sense?
> 
> If I'm right, adding a long-enough sleep between the change of RF and the start of the repair can make this error go away.

As far as I can see in the code, `ALTER KEYSPACE` only sends some schema mutations around, and the recomputation of replication maps happens asynchronously, so this race could be possible. However, it seems very unlikely, because — even without an explicit sleep — there is a sizeable time gap between the schema change and the repair call (logs taken from the original issue report):
node1:
```
INFO  2022-10-02 10:57:22,437 [shard 0] schema_tables - Altering keyspace ks
INFO  2022-10-02 10:57:22,443 [shard 0] schema_tables - Schema version changed to 0aa0ebcf-0ba3-3edf-976c-4ea38651bd7a
INFO  2022-10-02 10:57:23,966 [shard 0] repair - repair[4346c4a2-77ab-4490-bb21-1c600379c9ef]: starting user-requested repair for keyspace ks, repair id 1, options {{ trace -> false}, { jobThreads -> 1}, { incremental -> false}, { parallelism -> parallel}, { primaryRange -> false}}
```
node2:
```
INFO  2022-10-02 10:57:22,437 [shard 0] schema_tables - Altering keyspace ks
INFO  2022-10-02 10:57:22,443 [shard 0] schema_tables - Schema version changed to 0aa0ebcf-0ba3-3edf-976c-4ea38651bd7a
```

So the new keyspace metadata are propagated to the both nodes near-instantly, but repair only starts 1.5s later. This should be more than enough time to calculate new replication maps.

Also, why is only shard 1 skipping ranges due to "no followers", but shard 0 isn't? Does shard 0 correctly think that RF=2, while shard 1 is stuck thinking that RF=1?

**Comment by michoecho** [comment]
> The first rule of dtest - everything that is possible, will happen, and will waste 10 times the amount of time you could have spent on avoiding the problem in the first place :-(

Of course it deserves to be fixed, but I don't think it's the problem we're seeing here.

> By the way, I'm also not sure that the "schema version" messages you quoted is even relevant. We're talking here about the topology, not the schema, no?

`ALTER KEYSPACE` is a schema-modifying statement. This modification does change replication maps, but it's all is a part of the schema change. In particular, as far as I see from the code, effective replication maps (which the "repair neighbors" are taken from) are recalculated between the `Altering keyspace ks` message and the `Schema version changed` message. The fact that both messages happen before the start of repair seems to me to prove that this is not a race between the `ALTER KEYSPACE` and the start of repair. But this is the first time I'm reading the relevant code, so I might be missing something.

> If you can reproduce this issue (can you?),

I didn't attempt to.

**Comment by michoecho** [comment]
> Are you saying that the actual token ring (the list of tokens and who owns them) is written by the node doing the ALTER KEYSPACE as part of the original ALTER KEYSPACE operation?

No. The token ring (the "topology") and the replication strategy (the "schema") are orthogonal. The replication map is a function of both. ALTER KEYSPACE modifies the replication strategy, but not the token ring. IIUC gossip is responsible for topology updates, but not for schema updates, so ALTER KEYSPACE and gossip are unconnected.
That's my understanding anyway.

**Comment by nyh** [comment]
@michoecho I agree that re-setting RF does not change the "ring" itself (the **list** of tokens) but it does change the token ownership information (each node now knows it owns more of the token ranges, in fact - all of them), and I was wondering if this information is gossipped or just calculated independently on each node based on the old token ring and new RF?

**Comment by bhalevy** [comment]
the code path on the nodes in response to alter keyspace should go as follows:
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/db/schema_tables.cc#L1366
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L890
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L1226

Then:
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L891
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/migration_manager.cc#L451
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/storage_service.hh#L470
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/storage_service.cc#L3400
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/storage_service.cc#L3391-L3393
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/storage_service.cc#L3375-L3381

and eventually, from `mutate_token_metadata`:
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/storage_service.cc#L3368
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/service/storage_service.cc#L1498

Where we create the new effective replication map based on the replication strategy AND its configuration with the new RF.

Only then, repair should run and rebuild the now-missing token replicas (due to RF increase)

**Comment by bhalevy** [comment]
> @michoecho I agree that re-setting RF does not change the "ring" itself (the **list** of tokens) but it does change the token ownership information (each node now knows it owns more of the token ranges, in fact - all of them), and I was wondering if this information is gossipped or just calculated independently on each node based on the old token ring and new RF?

The information about the change is propagated via the schema update path.
The replication map is recalculated based on the ring (known as token_metadata), the replication strategy for the keyspace, and the replication strategy parameters (like the RF per DC for network topology strategy)

**Comment by michoecho** [comment]
> Then:

Wait. You say https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L891

but one line earlier there is:

https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L890
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L1226
https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/replica/database.cc#L1204

So I guess the ERM update happens twice?

**Comment by bhalevy** [comment]
> So I guess the ERM update happens twice?
The erm factory should find it in its `_effective_replication_maps` the 2nd time around.

**Comment by kbr-scylla** [comment]
> re-setting RF does not change the "ring" itself (the list of tokens) but it does change the token ownership information (each node now knows it owns more of the token ranges, in fact - all of them)

Don't confuse **owning** a token range vs **replicating** it.

Token ownership information is part of the token ring and it's not affected by the replication strategy (which RF is part of). Each token range has exactly one owner.

Replication strategy is part of the schema, not affected by the token ring.

To calculate replicas, you combine replication strategy with token ownership.

**Comment by Jadw1** [comment]
Another error happened in this test, but this time at the beginning of the test
https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2268/testReport/ttl_test/TestDistributedTTL/Tests___Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_/

```
self = <ttl_test.TestDistributedTTL object at 0x7f11098a8750>
enable_parallized_aggregation = 'false'

    @pytest.mark.next_gating
    @pytest.mark.dtest_debug
    @pytest.mark.parametrize("enable_parallized_aggregation", ['true', 'false'], ids=["parallel_aggregates_enabled", "parallel_aggregates_disabled"])
    def test_ttl_is_respected_on_repair(self, enable_parallized_aggregation):
        """ Test that ttl is respected on repair """
    
        self.prepare(jvm_args=['--logger-log-level', 'forward_service=trace',
                     '--enable-parallelized-aggregation', enable_parallized_aggregation])
        self.session1.execute("""
            ALTER KEYSPACE ks WITH REPLICATION =
            {'class' : 'SimpleStrategy', 'replication_factor' : 1};
        """)
        self.session1.execute("""
            INSERT INTO ttl_table (key, col1) VALUES (1, 1) USING TTL 5;
        """)
        self.session1.execute("""
            INSERT INTO ttl_table (key, col1) VALUES (2, 2) USING TTL 1000;
        """)
    
>       assert_all(
            self.session1,
            "SELECT * FROM ttl_table;",
            [[1, 1, None, None], [2, 2, None, None]]
        )
```
```
AssertionError: Expected [[1, 1, None, None], [2, 2, None, None]] from SELECT * FROM ttl_table;, but got [[1, 1, None, None]]
```

**Comment by bhalevy** [comment]
> Another error happened in this test, but this time at the beginning of the test https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2268/testReport/ttl_test/TestDistributedTTL/Tests___Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_/
> 
> ```
> self = <ttl_test.TestDistributedTTL object at 0x7f11098a8750>
> enable_parallized_aggregation = 'false'
> 
>     @pytest.mark.next_gating
>     @pytest.mark.dtest_debug
>     @pytest.mark.parametrize("enable_parallized_aggregation", ['true', 'false'], ids=["parallel_aggregates_enabled", "parallel_aggregates_disabled"])
>     def test_ttl_is_respected_on_repair(self, enable_parallized_aggregation):
>         """ Test that ttl is respected on repair """
>     
>         self.prepare(jvm_args=['--logger-log-level', 'forward_service=trace',
>                      '--enable-parallelized-aggregation', enable_parallized_aggregation])
>         self.session1.execute("""
>             ALTER KEYSPACE ks WITH REPLICATION =
>             {'class' : 'SimpleStrategy', 'replication_factor' : 1};
>         """)
>         self.session1.execute("""
>             INSERT INTO ttl_table (key, col1) VALUES (1, 1) USING TTL 5;
>         """)
>         self.session1.execute("""
>             INSERT INTO ttl_table (key, col1) VALUES (2, 2) USING TTL 1000;
>         """)
>     
> >       assert_all(
>             self.session1,
>             "SELECT * FROM ttl_table;",
>             [[1, 1, None, None], [2, 2, None, None]]
>         )
> ```
> 
> ```
> AssertionError: Expected [[1, 1, None, None], [2, 2, None, None]] from SELECT * FROM ttl_table;, but got [[1, 1, None, None]]
> ```

@Jadw1  This is a different issue.
It's not about wrong `count(*)` but rather unexpected result from `select *`.
Please look if there's another issue already opened for that failure
or open a separate issue for it otherwise.

**Comment by eliransin** [comment]
About this failure above, @nyh had a good point, maybe the schema hasn't been synced on the RF=1 yet so this select wrongly thought it can query all of the data locally. I suggest to add wait for a schema agreement after the change so it
is obvious that both nodes should be queried for the data.

**Comment by DoronArazii** [comment]
Pushing to 5.3 timeframe

**Comment by avikivity** [comment]
@mykaul let's prioritize, it's generating many failures in CI

**Comment by michoecho** [comment]
@avikivity There is a relevant active PR (from today) from @bhalevy (https://github.com/scylladb/scylla-dtest/pull/3086). IIUC the problem with this failure is that we can't reproduce it locally, and Benny's PR adds some changes to narrow it down.

**Comment by bhalevy** [comment]
@asias, after increasing RF, the test runs repair only on node1.
I was under the impression that repair is symmetric and so it would repair data on all peers the repair master syncs with.
But should the test run full repair on both (all) nodes?

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/47/artifact/logs-gating.release.2/dtest-gw6.log
```
06:34:45,957 783     ttl_test                       DEBUG    ttl_test.py         :765  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: SStable dump after writes with RF = 1
06:34:45,959 783     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=flush wait=True timeout=None
06:34:46,809 783     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node1: nodetool cmd=cfstats wait=True timeout=60
06:34:50,034 783     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-03-16T06:34:38.951917Z", "ttl" : 1000, "expires_at" : "2023-03-16T06:51:18Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
06:34:50,035 783     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=flush wait=True timeout=None
06:34:50,996 783     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=cfstats wait=True timeout=60
06:34:54,256 783     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-03-16T06:34:38.949088Z", "ttl" : 5, "expires_at" : "2023-03-16T06:34:43Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
```

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/47/artifact/logs-gating.release.2/1678948539931_ttl_test.py%3A%3ATestDistributedTTL%3A%3Atest_ttl_is_respected_on_repair%5Bparallel_aggregates_disabled%5D/node1.log
```
INFO  2023-03-16 06:35:22,236 [shard 0] schema_tables - Altering keyspace ks
INFO  2023-03-16 06:35:22,243 [shard 0] schema_tables - Schema version changed to fba21852-02d5-3f97-9d0f-db8415ac0426
INFO  2023-03-16 06:35:23,832 [shard 0] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: starting user-requested repair for keyspace ks, repair id 1, options {{trace -> false}, {jobThreads -> 1}, {incremental -> false}, {parallelism -> parallel}, {primaryRange -> false}}
INFO  2023-03-16 06:35:23,832 [shard 0] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: Skipped sending repair_flush_hints_batchlog to nodes=[127.0.12.2, 127.0.12.1]
INFO  2023-03-16 06:35:23,832 [shard 0] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: Started to repair 1 out of 1 tables in keyspace=ks, table=ttl_table, table_id=9f17b210-c3c4-11ed-b4d3-c38932084eeb, repair_reason=repair
INFO  2023-03-16 06:35:23,832 [shard 1] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: Started to repair 1 out of 1 tables in keyspace=ks, table=ttl_table, table_id=9f17b210-c3c4-11ed-b4d3-c38932084eeb, repair_reason=repair
...
INFO  2023-03-16 06:35:23,934 [shard 1] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: stats: repair_reason=repair, keyspace=ks, tables={ttl_table}, ranges_nr=513, round_nr=257, round_nr_fast_path_already_synced=257, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=0, rpc_call_nr=1285, tx_hashes_nr=0, rx_hashes_nr=0, duration=0.10202415 seconds, tx_row_nr=0, rx_row_nr=0, tx_row_bytes=0, rx_row_bytes=0, row_from_disk_bytes={{127.0.12.1, 0}, {127.0.12.2, 0}}, row_from_disk_nr={{127.0.12.1, 0}, {127.0.12.2, 0}}, row_from_disk_bytes_per_sec={{127.0.12.1, 0}, {127.0.12.2, 0}} MiB/s, row_from_disk_rows_per_sec={{127.0.12.1, 0}, {127.0.12.2, 0}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={}
INFO  2023-03-16 06:35:23,934 [shard 1] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: completed successfully, keyspace=ks
INFO  2023-03-16 06:35:24,056 [shard 0] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: stats: repair_reason=repair, keyspace=ks, tables={ttl_table}, ranges_nr=513, round_nr=514, round_nr_fast_path_already_synced=513, round_nr_fast_path_same_combined_hashes=0, round_nr_slow_path=1, rpc_call_nr=2568, tx_hashes_nr=0, rx_hashes_nr=1, duration=0.22416809 seconds, tx_row_nr=0, rx_row_nr=1, tx_row_bytes=0, rx_row_bytes=172, row_from_disk_bytes={{127.0.12.1, 0}, {127.0.12.2, 172}}, row_from_disk_nr={{127.0.12.1, 0}, {127.0.12.2, 1}}, row_from_disk_bytes_per_sec={{127.0.12.1, 0}, {127.0.12.2, 0.000731737}} MiB/s, row_from_disk_rows_per_sec={{127.0.12.1, 0}, {127.0.12.2, 4.46094}} Rows/s, tx_row_nr_peer={}, rx_row_nr_peer={{127.0.12.2, 1}}
INFO  2023-03-16 06:35:24,056 [shard 0] repair - repair[5cff8b62-acf2-4848-afc5-59e7dc1b9d50]: completed successfully, keyspace=ks
```

So thre's only `rx_row_nr=1` on shard 0

And then, https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/47/artifact/logs-gating.release.2/dtest-gw6.log
shows the missing partition on node2
```
06:35:30,325 783     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-03-16T06:34:38.951917Z", "ttl" : 1000, "expires_at" : "2023-03-16T06:51:18Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-03-16T06:34:38.949088Z", "ttl" : 5, "expires_at" : "2023-03-16T06:34:43Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
06:35:30,325 783     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=flush wait=True timeout=None
06:35:31,107 783     ccm                            DEBUG    cluster.py          :711  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: node2: nodetool cmd=cfstats wait=True timeout=60
06:35:33,815 783     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_disabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-03-16T06:34:38.949088Z", "ttl" : 5, "expires_at" : "2023-03-16T06:34:43Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
```

**Comment by bhalevy** [comment]
Sigh, I reviewed the test code again and my explanations are off a bit since the test starts the 2-node cluster before writing the 2 rows.  Those are written immediately after `ALTER KEYSPACE ks WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 1};` so I think it's possible they race with the schema change and that's why the first row (with ttl=5) is written to both replicas while the 2nd row (with ttl=1000) is written only once.

I'll continue to fix the test to take all this stuff into account and to run repair on both nodes (as the [doc](https://docs.scylladb.com/stable/operating-scylla/nodetool-commands/repair.html) says)

**Comment by benipeled** [comment]
Seen on 5.1 - https://jenkins.scylladb.com/job/scylla-5.1/job/dtest-release/75/testReport/junit/ttl_test/TestDistributedTTL/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split004___test_ttl_is_respected_on_repair/

Please backport the required

**Comment by fruch** [comment]
> Seen on 5.1 - https://jenkins.scylladb.com/job/scylla-5.1/job/dtest-release/75/testReport/junit/ttl_test/TestDistributedTTL/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split004___test_ttl_is_respected_on_repair/
> 
> Please backport the required

https://github.com/scylladb/scylla-dtest/pull/3086 was backport
